File: ruby/rbsearch/searcher.rb
# frozen_string_literal: true

require 'find'
require 'pathname'
require 'set'
require_relative 'common.rb'
require_relative 'filetypes.rb'
require_relative 'fileutil.rb'
require_relative 'searcherror.rb'
require_relative 'searchfile.rb'
require_relative 'searchresult.rb'
require_relative 'searchresultformatter.rb'

# Searcher - finds files to search and searches them according to settings
class Searcher
  attr_reader :settings
  attr_reader :results

  def initialize(settings)
    @settings = settings
    validate_settings
    @filetypes = FileTypes.new
    @results = []
  end

  def search_dir?(dirname)
    path_elems = dirname.split(File::SEPARATOR) - FileUtil.dot_dirs
    if @settings.excludehidden && path_elems.any? { |p| FileUtil.hidden?(p) }
      return false
    end
    if !@settings.in_dirpatterns.empty? &&
       !any_matches_any_pattern(path_elems, @settings.in_dirpatterns)
      return false
    end
    if !@settings.out_dirpatterns.empty? &&
       any_matches_any_pattern(path_elems, @settings.out_dirpatterns)
      return false
    end
    true
  end

  def search_file?(filename)
    ext = FileUtil.get_extension(filename)
    if !@settings.in_extensions.empty? &&
       !@settings.in_extensions.include?(ext)
      return false
    end
    if !@settings.out_extensions.empty? &&
       @settings.out_extensions.include?(ext)
      return false
    end
    if !@settings.in_filepatterns.empty? &&
       !matches_any_pattern(filename, @settings.in_filepatterns)
      return false
    end
    if !@settings.out_filepatterns.empty? &&
       matches_any_pattern(filename, @settings.out_filepatterns)
      return false
    end
    filetype = @filetypes.get_filetype(filename)
    if !@settings.in_filetypes.empty? &&
       !@settings.in_filetypes.include?(filetype)
      return false
    end
    if !@settings.out_filetypes.empty? &&
       @settings.out_filetypes.include?(filetype)
      return false
    end
    true
  end

  def archive_search_file?(filename)
    ext = FileUtil.get_extension(filename)
    if !@settings.in_archiveextensions.empty? &&
       !@settings.in_archiveextensions.include?(ext)
      return false
    end
    if !@settings.out_archiveextensions.empty? &&
       @settings.out_archiveextensions.include?(ext)
      return false
    end
    if !@settings.in_archivefilepatterns.empty? &&
       !matches_any_pattern(filename, @settings.in_archivefilepatterns)
      return false
    end
    if !@settings.out_archivefilepatterns.empty? &&
       matches_any_pattern(filename, @settings.out_archivefilepatterns)
      return false
    end
    true
  end

  def filter_file?(filename)
    if @settings.excludehidden && FileUtil.hidden?(filename)
      return false
    end
    if @filetypes.archive_file?(filename)
      return @settings.searcharchives && archive_search_file?(filename)
    end
    !@settings.archivesonly && search_file?(filename)
  end

  def search
    # get the searchfiles
    searchfiles = get_search_files.sort_by(&:relativepath)
    if @settings.verbose
      searchdirs = searchfiles.map(&:path).uniq.sort
      log("\nDirectories to be searched (#{searchdirs.size}):")
      searchdirs.each do |d|
        log(d)
      end
      log("\nFiles to be searched (#{searchfiles.size}):")
      searchfiles.each do |sf|
        log(sf.to_s)
      end
      log("\n")
    end
    searchfiles.each do |sf|
      search_file(sf)
    end
  end

  def search_multiline_string(str)
    lines_before = []
    lines_after = []
    results = []
    new_line_indices = get_new_line_indices(str)
    start_line_indices = [0] + new_line_indices.map { |n| n + 1 }
    end_line_indices = new_line_indices + [str.length - 1]
    @settings.searchpatterns.each do |p|
      m = p.match(str)
      stop = false
      while m && !stop
        m_line_start_index = 0
        m_line_end_index = str.length - 1
        before_start_indices = start_line_indices.take_while { |i| i <= m.begin(0) }
        before_line_count = 0
        if before_start_indices
          m_line_start_index = before_start_indices.last
          before_line_count = before_start_indices.size - 1
        end
        m_line_end_index = end_line_indices[start_line_indices.index(m_line_start_index)]
        after_start_indices = start_line_indices.drop_while { |i| i <= m.begin(0) }
        line = str[m_line_start_index..m_line_end_index]
        if @settings.linesbefore && before_line_count
          before_start_indices = before_start_indices[0, before_start_indices.size - 1].last(@settings.linesbefore)
          lines_before = get_lines_before(str, before_start_indices,
                                          start_line_indices, end_line_indices)
        end
        if @settings.linesafter && after_start_indices
          after_start_indices = after_start_indices[0, @settings.linesafter]
          lines_after = get_lines_after(str, after_start_indices,
                                        start_line_indices, end_line_indices)
        end
        match_start_index = m.begin(0) - m_line_start_index
        match_end_index = m.end(0) - m_line_start_index
        if (lines_before.length == 0 || lines_before_match(lines_before)) &&
           (lines_after.length == 0 || lines_after_match(lines_after))
          results.push(SearchResult.new(
            p,
            nil,
            before_line_count + 1,
            match_start_index + 1,
            match_end_index + 1,
            line,
            lines_before,
            lines_after
          ))
          stop = true if @settings.firstmatch
        end
        m = p.match(str, m_line_start_index + match_end_index)
      end
    end
    results
  end

  def search_line_iterator(lines)
    linenum = 0
    lines_before = []
    lines_after = []
    pattern_matches = {}
    results = []
    loop do
      line =
        if lines_after.empty?
          begin
            lines.next
          rescue StopIteration
            return results
          end
        else
          lines_after.shift
        end
      linenum += 1
      if @settings.linesafter
        while lines_after.length < @settings.linesafter
          begin
            lines_after.push(lines.next)
          rescue StopIteration
            break
          end
        end
      end
      @settings.searchpatterns.each do |p|
        search_line = true
        if @settings.firstmatch && pattern_matches.include?(p)
          search_line = false
        end
        pos = 0
        while search_line && pos < line.length
          begin
            m = p.match(line, pos)
            if m
              if (!lines_before.empty? && !lines_before_match(lines_before)) ||
                 (!lines_after.empty? && !lines_after_match(lines_after))
                search_line = false
              else
                results.push(SearchResult.new(
                  p,
                  nil,
                  linenum,
                  m.begin(0) + 1,
                  m.end(0) + 1,
                  line,
                  [].concat(lines_before),
                  [].concat(lines_after)
                ))
                pos = m.end(0) + 1
                pattern_matches[p] = 1
              end
            else
              search_line = false
            end
          rescue StandardError => e
            raise SearchError, e
          end
        end
      end
      if @settings.linesbefore
        lines_before.shift if lines_before.size == @settings.linesbefore
        lines_before.push(line) if lines_before.size < @settings.linesbefore
      end
    end
  end

  def print_results
    formatter = SearchResultFormatter.new(@settings)
    log("Search results (#{@results.size}):")
    @results.each do |r|
      # print_result(r)
      log(formatter.format(r))
    end
  end

  def print_result(search_result)
    s = ''
    s += "#{search_result.pattern}: " if @settings.searchpatterns.size > 1
    s += search_result.to_s
    log(s)
  end

  def get_matching_dirs
    @results.map { |r| r.file.path }.uniq.sort
  end

  def get_matching_files
    @results.map { |r| r.file.to_s }.uniq.sort
  end

  def get_matching_lines
    lines = @results.map { |r| r.line.strip }.sort { |l1, l2| l1.upcase <=> l2.upcase }
    lines.uniq! if @settings.uniquelines
    lines
  end

  private

  def validate_settings
    raise SearchError, 'Startpath not defined' unless @settings.startpath
    raise SearchError, 'Startpath not found' unless Pathname.new(@settings.startpath).exist?
    raise SearchError, 'Startpath not readable' unless File.readable?(@settings.startpath)
    raise SearchError, 'No search patterns defined' if @settings.searchpatterns.empty?
    raise SearchError, 'Invalid linesbefore' if @settings.linesbefore < 0
    raise SearchError, 'Invalid linesafter' if @settings.linesafter < 0
    raise SearchError, 'Invalid maxlinelength' if @settings.maxlinelength < 0
    begin
      _enc = Encoding.find(@settings.textfileencoding)
    rescue StandardError
      raise SearchError, "Invalid encoding: #{@settings.textfileencoding}"
    end
  end

  def matches_any_pattern(str, pattern_set)
    pattern_set.any? { |p| p.match(str) }
  end

  def any_matches_any_pattern(str_list, pattern_set)
    str_list.each do |s|
      return true if matches_any_pattern(s, pattern_set)
    end
    false
  end

  def file_to_searchfile(f)
    d = File.dirname(f) || '.'
    filename = File.basename(f)
    filetype = @filetypes.get_filetype(filename)
    SearchFile.new(d, filename, filetype)
  end

  def get_search_files
    searchfiles = []
    if FileTest.directory?(@settings.startpath)
      if @settings.recursive
        Find.find(@settings.startpath) do |f|
          if FileTest.directory?(f)
            Find.prune unless search_dir?(f)
          elsif filter_file?(f)
            searchfile = file_to_searchfile(f)
            searchfiles.push(searchfile)
          end
        end
      else
        Find.find(@settings.startpath) do |f|
          if FileTest.directory?(f)
            Find.prune
          elsif filter_file?(f)
            searchfile = file_to_searchfile(f)
            searchfiles.push(searchfile)
          end
        end
      end
    elsif FileTest.file?(@settings.startpath)
      searchfile = file_to_searchfile(@settings.startpath)
      searchfiles.push(searchfile)
    end
    searchfiles
  end

  def search_file(sf)
    unless @filetypes.searchable_file?(sf.filename)
      if @settings.verbose || @settings.debug
        log("Skipping unsearchable file: #{sf}")
        return 0
      end
    end
    case sf.filetype
    when FileType::CODE, FileType::TEXT, FileType::XML
      search_text_file(sf)
    when FileType::BINARY
      search_binary_file(sf)
    else
      log("Searching currently unsupported for FileType #{sf.filetype}")
    end
  end

  def search_binary_file(sf)
    f = File.open(sf.relativepath, 'rb')
    contents = f.read
    results = search_binary_string(contents)
    results.each do |r|
      r.file = sf
      add_search_result(r)
    end
  rescue StandardError => e
    raise SearchError, e
  ensure
    f&.close
  end

  def search_binary_string(binstr)
    results = []
    @settings.searchpatterns.each do |p|
      pos = 0
      m = p.match(binstr, pos)
      stop = false
      while m && !stop
        results.push(SearchResult.new(
          p,
          nil,
          0,
          m.begin(0) + 1,
          m.end(0) + 1,
          nil
        ))
        pos = m.end(0) + 1
        m = p.match(binstr, pos)
        stop = true if @settings.firstmatch
      end
    end
    results
  end

  def search_text_file(sf)
    log("Searching text file #{sf}") if @settings.debug
    if @settings.multilinesearch
      search_text_file_contents(sf)
    else
      search_text_file_lines(sf)
    end
  end

  def get_line_count(s)
    s.scan(/(\r\n|\n)/m).size
  end

  def search_text_file_contents(sf)
    f = File.open(sf.relativepath, mode: "r:#{@settings.textfileencoding}")
    contents = f.read
    results = search_multiline_string(contents)
    results.each do |r|
      r.file = sf
      add_search_result(r)
    end
  rescue StandardError => e
    raise SearchError, "#{e} (file: #{sf})"
  ensure
    f&.close
  end

  def get_new_line_indices(s)
    indices = []
    i = 0
    s.chars.each do |c|
      indices.push(i) if c == "\n"
      i += 1
    end
    indices
  end

  def get_lines_at_indices(s, at_indices, start_line_indices, end_line_indices)
    lines = []
    at_indices.each do |i|
      line = s[i..end_line_indices[start_line_indices.index(i)]]
      lines.push(line)
    end
    lines
  end

  def get_lines_before(s, before_start_indices, start_line_indices, end_line_indices)
    get_lines_at_indices(s, before_start_indices, start_line_indices, end_line_indices)
  end

  def get_lines_after(s, after_start_indices, start_line_indices, end_line_indices)
    get_lines_at_indices(s, after_start_indices, start_line_indices, end_line_indices)
  end

  def lines_match(lines, in_patterns, out_patterns)
    ((in_patterns.empty? || any_matches_any_pattern(lines, in_patterns)) &&
    (out_patterns.empty? || !any_matches_any_pattern(lines, out_patterns)))
  end

  def lines_before_match(lines_before)
    lines_match(lines_before, @settings.in_linesbeforepatterns,
      @settings.out_linesbeforepatterns)
  end

  def do_lines_after_or_until
    !@settings.linesaftertopatterns.empty? ||
      !@settings.linesafteruntilpatterns.empty?
  end

  def do_lines_after
    @settings.linesafter || do_lines_after_or_until
  end

  def lines_after_match(lines_after)
    return true if do_lines_after_or_until
    lines_match(lines_after, @settings.in_linesafterpatterns,
      @settings.out_linesafterpatterns)
  end

  def search_text_file_lines(sf)
    f = File.open(sf.relativepath, mode: "r:#{@settings.textfileencoding}")
    line_iterator = f.each_line
    results = search_line_iterator(line_iterator)
    results.each do |r|
      r.file = sf
      add_search_result(r)
    end
  rescue StandardError => e
    raise SearchError, "#{e} (file: #{sf})"
  ensure
    f&.close
  end

  def add_search_result(search_result)
    @results.push(search_result)
  end
end

File: node/nodesearch/searcher.js
/*
 * searcher.js
 *
 * performs the searching based on the given SearchSettings instance
 */

var assert = require('assert');
var fs = require('fs');
var path = require('path');

var common = require('./common.js');
var FileType = require('./filetype.js').FileType;
var FileTypes = require('./filetypes.js').FileTypes;
var FileUtil = require('./fileutil.js').FileUtil;
var SearchResult = require('./searchresult.js').SearchResult;

function Searcher(settings) {
    var self = this;
    var _settings = settings;
    var _filetypes = new FileTypes();
    var _timers = {};
    var _totalElapsed = 0;
    self.results = [];

    var validateSettings = function () {
        assert.ok(_settings.startPath, 'Startpath not defined');
        assert.ok(fs.existsSync(_settings.startPath), 'Startpath not found');
        assert.ok(isSearchDir(_settings.startPath), 'Startpath does not match search settings');
        assert.ok(_settings.searchPatterns.length, 'No search patterns specified');
    };

    var matchesAnyElement = function (s, elements) {
        return elements.indexOf(s) > -1;
    };

    var matchesAnyPattern = function (s, patterns) {
        for (var p in patterns) {
            var pattern = patterns[p];
            var match = pattern.exec(s);
            if (match) {
                return true;
            }
        }
        return false;
    };

    var anyMatchesAnyPattern = function (ss, patterns) {
        for (var i in ss) {
            if (matchesAnyPattern(ss[i], patterns)) {
                return true;
            }
        }
        return false;
    };

    var isSearchDir = function (dir) {
        var pathElems = dir.split(path.sep);
        if (_settings.excludeHidden) {
            for (var p in pathElems) {
                if (!matchesAnyElement(pathElems[p], ['.','..']) &&
                    FileUtil.isHidden(pathElems[p])) {
                    return false;
                }
            }
        }
        if (_settings.inDirPatterns.length && !matchesAnyPattern(dir,
            _settings.inDirPatterns)) {
            return false;
        }
        if (_settings.outDirPatterns.length && matchesAnyPattern(dir,
            _settings.outDirPatterns)) {
            return false;
        }
        return true;
    };
    // can validate now that isSearchDir is defined
    validateSettings();

    var isSearchFile = function (file) {
        if (FileUtil.isHidden(file) && _settings.excludeHidden) {
            return false;
        }
        var ext = FileUtil.getExtension(file);
        if (_settings.inExtensions.length &&
            !matchesAnyElement(ext, _settings.inExtensions)) {
            return false;
        }
        if (_settings.outExtensions.length &&
            matchesAnyElement(ext, _settings.outExtensions)) {
            return false;
        }
        if (_settings.inFilePatterns.length &&
            !matchesAnyPattern(file, _settings.inFilePatterns)) {
            return false;
        }
        if (_settings.outFilePatterns.length &&
            matchesAnyPattern(file, _settings.outFilePatterns)) {
            return false;
        }
        return true;
    };

    var isArchiveSearchFile = function (file) {
        if (FileUtil.isHidden(file) && _settings.excludeHidden) {
            return false;
        }
        var ext = FileUtil.getExtension(file);
        if (_settings.inArchiveExtensions.length &&
            !matchesAnyElement(ext, _settings.inArchiveExtensions)) {
            return false;
        }
        if (_settings.outArchiveExtensions.length &&
            matchesAnyElement(ext, _settings.outArchiveExtensions)) {
            return false;
        }
        if (_settings.inArchiveFilePatterns.length &&
            !matchesAnyPattern(file, _settings.inArchiveFilePatterns)) {
            return false;
        }
        if (_settings.outArchiveFilePatterns.length &&
            matchesAnyPattern(file, _settings.outArchiveFilePatterns)) {
            return false;
        }
        return true;
    };

    var getSearchDirs = function (startPath) {
        var searchDirs = [];
        var stats = fs.statSync(startPath);
        if (stats.isDirectory()) {
            if (_settings.debug) {
                common.log("Getting list of directories to search under {0}".format(startPath));
            }
            searchDirs.push(startPath);
            if (_settings.recursive) {
                searchDirs.push.apply(searchDirs, recGetSearchDirs(startPath));
            }
        } else if (stats.isFile()) {
            var d = path.dirname(startPath);
            if (!d) d = ".";
            searchDirs.push(d);
        }
        return searchDirs;
    };

    var handleFsError = function (err) {
        if (err.errno === 34 && err.code === "ENOENT") {
            // this error seems to occur when the file is a soft link
            // to a non-existent file
        } else {
            common.log(err);
            process.exit(1);
        }
    };

    var getSubDirs = function (dir) {
        var subDirs = [];
        var childItems = fs.readdirSync(dir);
        for (var c in childItems) {
            var filepath = path.join(dir, childItems[c]);
            try {
                var stats = fs.statSync(filepath);
                if (stats.isDirectory()) {
                    subDirs.push(filepath);
                }
            } catch (err) {
                handleFsError(err);
            }
        }
        return subDirs;
    };

    var recGetSearchDirs = function (currentDir) {
        var searchDirs = [];
        var subDirs = getSubDirs(currentDir);
        for (var d in subDirs) {
            if (isSearchDir(subDirs[d])) {
                searchDirs.push(subDirs[d]);
            }
            searchDirs.push.apply(searchDirs, recGetSearchDirs(subDirs[d]));
        }
        return searchDirs;
    };

    var getFilesForDirectory = function (dir) {
        var files = [];
        var childItems = fs.readdirSync(dir);
        for (var c in childItems) {
            var filepath = path.join(dir, childItems[c]);
            try {
                var stats = fs.statSync(filepath);
                if (stats.isFile()) {
                    files.push(filepath);
                }
            } catch (err) {
                handleFsError(err);
            }
        }
        return files;
    };

    var filterFile = function (f) {
        if (_filetypes.isArchiveFile(f) && _settings.searchArchives &&
            isArchiveSearchFile(f))
            return true;
        if (!_settings.archivesOnly && isSearchFile(f))
            return true;
        return false;
    };

    var getSearchFilesForDirectory = function (dir) {
        var searchFiles = [];
        var dirFiles = getFilesForDirectory(dir);
        for (var d in dirFiles) {
            var f = dirFiles[d];
            if (filterFile(f)) {
                searchFiles.push(f);
            }
        }
        return searchFiles;
    };

    var getSearchFiles = function (searchDirs) {
        var searchFiles = [];
        var stats = fs.statSync(_settings.startPath);
        if (stats.isDirectory()) {
            for (var d in searchDirs) {
                searchFiles.push.apply(searchFiles, getSearchFilesForDirectory(searchDirs[d]));
            }
        } else if (stats.isFile()) {
            searchFiles.push(_settings.startPath);
        }
        return searchFiles;
    };

    var addTimer = function (name, action) {
        _timers[name + ":" + action] = new Date();
    };

    var startTimer = function (name) {
        addTimer(name, "start");
    };

    var stopTimer = function (name) {
        addTimer(name, "stop");
        addElapsed(name);
    };

    var addElapsed = function (name) {
        _totalElapsed += getElapsed(name);
    };

    var getElapsed = function (name) {
        var start = _timers[name+":start"];
        var stop = _timers[name+":stop"];
        var elapsed = stop - start;
        return elapsed;
    };

    var printElapsed = function (name) {
        var elapsed = getElapsed(name);
        var msg = "Elapsed time for {0}: {1} ms";
        common.log(msg.format(name, elapsed));
    };

    var printTotalElapsed = function () {
        var msg = "Total elapsed time: {0} ms";
        common.log(msg.format(_totalElapsed));
    };

    self.search = function () {
        if (_settings.verbose)
            common.log("Search initiated");

        // get the search dirs
        if (_settings.doTiming)
            startTimer("GetSearchDirs");
        var dirs = [];
        dirs.push.apply(dirs, getSearchDirs(_settings.startPath));
        if (_settings.doTiming) {
            stopTimer("GetSearchDirs");
            if (_settings.printResults)
                printElapsed("GetSearchDirs");
        }
        if (_settings.verbose) {
            common.log("\nDirectories to be searched ({0}):".format(dirs.length));
            for (var d in dirs) {
                common.log(dirs[d]);
            }
        }

        // get the search files
        if (_settings.doTiming)
            startTimer("GetSearchFiles");
        var files = getSearchFiles(dirs);
        if (_settings.doTiming) {
            stopTimer("GetSearchFiles");
            if (_settings.printResults)
                printElapsed("GetSearchFiles");
        }
        if (_settings.verbose) {
            common.log("\nFiles to be searched ({0}):".format(files.length));
            for (var f in files) {
                common.log(files[f]);
            }
            common.log("");
        }

        if (_settings.doTiming)
            startTimer("SearchFiles");
        for (var i in files) {
            searchFile(files[i]);
        }
        if (_settings.doTiming) {
            stopTimer("SearchFiles");
            if (_settings.printResults) {
                printElapsed("SearchFiles");
                printTotalElapsed();
            }
        }

        if (_settings.verbose)
            common.log("Search complete.");
    };

    var searchFile = function (filepath) {
        var filetype = _filetypes.getFileType(filepath);
        if (filetype === FileType.TEXT) {
            searchTextFile(filepath);
        } else if (filetype === FileType.BINARY) {
            searchBinaryFile(filepath);
        }
    };

    var searchBinaryFile = function (filepath) {
        if (_settings.verbose) {
            common.log('Searching binary file: "{0}"'.format(filepath));
        }
        var contents = fs.readFileSync(filepath).toString();
        var pattern = '';
        for (var p in _settings.searchPatterns) {
            pattern = _settings.searchPatterns[p];
            var match = pattern.exec(contents);
            if (match) {
                addSearchResult(new SearchResult(pattern, filepath, 0, 0, 0,
                    null, [], []));
            }
        }
    };

    var searchTextFile = function (filepath) {
        if (_settings.verbose) {
            common.log('Searching text file {0}'.format(filepath));
        }
        if (_settings.multilineSearch)
            searchTextFileContents(filepath);
        else
            searchTextFileLines(filepath);
    };

    var getLineCount = function (contents) {
        var lineCount = 0;
        if (contents) {
            var matches = contents.match(/(\r?\n)/g);
            if (matches) lineCount = matches.length;
        }
        return lineCount;
    };

    var searchTextFileContents = function (filepath) {
        var contents = fs.readFileSync(filepath).toString();
        var results = self.searchMultiLineString(contents);
        for (var i in results) {
            var r = results[i];
            var resultWithFilepath =
                new SearchResult(r.pattern, filepath, r.linenum,
                        r.matchStartIndex, r.matchEndIndex, r.line,
                        r.linesBefore, r.linesAfter);
            addSearchResult(resultWithFilepath);
        }
    };

    var getNewLineIndices = function (s) {
        var indices = [];
        for (var i = 0; i < s.length; i++) {
            if (s.charAt(i) == "\n") {
                indices.push(i);
            }
        }
        return indices;
    };

    var getLinesAtIndices = function(s, atIndices, startLineIndices, endLineIndices) {
        if (atIndices.length === 0)
            return [];
        var lines = [];
        for (var a in atIndices) {
            var i = atIndices[a];
            var line = s.substring(i, endLineIndices[startLineIndices.indexOf(i)]);
            lines.push(line);
        }
        return lines;
    };

    var getLinesBefore = function(s, beforeStartIndices, startLineIndices, endLineIndices) {
        return getLinesAtIndices(s, beforeStartIndices, startLineIndices, endLineIndices);
    };

    var getLinesAfter = function(s, afterStartIndices, startLineIndices, endLineIndices) {
        return getLinesAtIndices(s, afterStartIndices, startLineIndices, endLineIndices);
    };

    var lessOrEqual = function(i) { return i <= match.index; };
    var greaterThan = function(i) { return i > match.index; };

    self.searchMultiLineString = function (s) {
        var patternResults = {};
        var linesBefore = [];
        var linesAfter = [];
        var results = [];
        var newLineIndices = getNewLineIndices(s);
        var plusOne = function(i) { return i+1; };
        var startLineIndices = [0].concat(newLineIndices.map(plusOne));
        var endLineIndices = newLineIndices.concat([s.length - 1]);
        for (var p in _settings.searchPatterns) {
            var pattern = new RegExp(_settings.searchPatterns[p].source, "g");
            if (_settings.firstMatch && pattern in patternResults)
                continue;
            var match = pattern.exec(s);
            while (match) {
                var lineStartIndex = 0;
                var lineEndIndex = s.length - 1;
                var beforeLineCount = 0;
                var beforeStartIndices = startLineIndices.filter(lessOrEqual);
                if (beforeStartIndices.length > 0) {
                    lineStartIndex = beforeStartIndices.pop();
                    beforeLineCount = beforeStartIndices.length;
                    if (beforeStartIndices.length > _settings.linesBefore) {
                        beforeStartIndices = beforeStartIndices.slice(
                            beforeStartIndices.length - _settings.linesBefore);
                    }
                }
                lineEndIndex = endLineIndices[startLineIndices.indexOf(lineStartIndex)];
                line = s.substring(lineStartIndex, lineEndIndex);
                if (_settings.linesBefore && beforeLineCount) {
                    linesBefore = getLinesBefore(s, beforeStartIndices,
                        startLineIndices, endLineIndices);
                }
                if (_settings.linesAfter) {
                    var afterStartIndices = startLineIndices.filter(greaterThan);
                    if (afterStartIndices.length > _settings.linesAfter) {
                        afterStartIndices = afterStartIndices.slice(0,
                            _settings.linesAfter);
                    }
                    linesAfter = getLinesAfter(s, afterStartIndices,
                        startLineIndices, endLineIndices);
                }
                var matchStartIndex = match.index - lineStartIndex + 1;
                var matchEndIndex = pattern.lastIndex - lineStartIndex + 1;
                if ((_settings.linesBefore === 0 || linesBeforeMatch(linesBefore)) &&
                    (_settings.linesAfter === 0 || linesAfterMatch(linesAfter))) {
                    var searchResult = new SearchResult(
                        pattern,
                        '',
                        beforeLineCount+1,
                        matchStartIndex,
                        matchEndIndex,
                        line,
                        [].concat(linesBefore),
                        [].concat(linesAfter));
                    results.push(searchResult);
                    if (!(pattern in patternResults))
                        patternResults[pattern] = [];
                    patternResults[pattern].push(searchResult);
                }
                match = pattern.exec(s);
            }
        }
        return results;
    };

    var linesMatch = function(lines, inPatterns, outPatterns) {
        return ((inPatterns.length === 0 || anyMatchesAnyPattern(lines, inPatterns)) &&
               (outPatterns.length === 0 || ! anyMatchesAnyPattern(lines, outPatterns)));
    };

    var linesBeforeMatch = function(linesBefore) {
        return linesMatch(linesBefore, _settings.inLinesBeforePatterns,
            _settings.outLinesBeforePatterns);
    };

    var linesAfterMatch = function(linesAfter) {
        return linesMatch(linesAfter, _settings.inLinesAfterPatterns,
            _settings.outLinesAfterPatterns);
    };

    var searchTextFileLines = function (filepath) {
        var contents = fs.readFileSync(filepath).toString();
        var lines = contents.toString().split(/\r?\n/);
        var results = searchLines(lines);
        for (var i in results) {
            var r = results[i];
            var resultWithFilepath =
                new SearchResult(r.pattern, filepath, r.linenum,
                        r.matchStartIndex, r.matchEndIndex, r.line,
                        r.linesBefore, r.linesAfter);
            addSearchResult(resultWithFilepath);
        }
    };

    // return results so that filepath can be added to them
    var searchLines = function (lines) {
        var linenum = 0;
        var pattern;
        var linesBefore = [];
        var linesAfter = [];
        var results = [];
        while (true) {
            var line = "";
            if (linesAfter.length > 0) {
                line = linesAfter.shift();
            } else if (lines.length > 0) {
                line = lines.shift();
            } else {
                return results;
            }
            linenum += 1;
            if (_settings.linesAfter > 0) {
                while (linesAfter.length < _settings.linesAfter && lines.length > 0) {
                    linesAfter.push(lines.shift());
                }
            }
            for (var p in _settings.searchPatterns) {
                pattern = new RegExp(_settings.searchPatterns[p].source, "g");
                var match = pattern.exec(line);
                while (match) {
                    if ((_settings.linesBefore === 0 || linesBeforeMatch(linesBefore)) &&
                        (_settings.linesAfter === 0 || linesAfterMatch(linesAfter))) {
                        results.push(new SearchResult(
                            pattern,
                            '',
                            linenum,
                            match.index+1,
                            pattern.lastIndex+1,
                            line,
                            [].concat(linesBefore),
                            [].concat(linesAfter)));
                        if (_settings.firstMatch) {
                            break;
                        }
                    }
                    match = pattern.exec(line);
                }
            }
            if (_settings.linesBefore > 0) {
                if (linesBefore.length == _settings.linesBefore)
                    linesBefore.shift();
                if (linesBefore.length < _settings.linesBefore)
                    linesBefore.push(line);
            }
        }
    };

    var addSearchResult = function (result) {
        self.results.push(result);
    };

    self.getMatchingDirs = function () {
        var dirs = [];
        for (var r in self.results) {
            var result = self.results[r];
            dirs.push(path.dirname(result.filename));
        }
        return common.setFromArray(dirs);
    };

    self.printMatchingDirs = function () {
        var dirs = self.getMatchingDirs();
        common.log("\nDirectories with matches ({0}):".format(dirs.length));
        for (var d in dirs) {
            common.log(dirs[d]);
        }
    };

    self.getMatchingFiles = function () {
        var files = [];
        for (var r in self.results) {
            var result = self.results[r];
            files.push(result.filename);
        }
        return common.setFromArray(files);
    };

    self.printMatchingFiles = function () {
        var files = self.getMatchingFiles();
        common.log("\nFiles with matches ({0}):".format(files.length));
        for (var f in files) {
            common.log(files[f]);
        }
    };

    self.getMatchingLines = function () {
        var lines = [];
        for (var r in self.results) {
            var result = self.results[r];
            if (result.linenum)
                lines.push(result.line.trim());
        }
        if (_settings.uniqueLines)
            lines = common.setFromArray(lines);
        lines.sort();
        return lines;
    };

    self.printMatchingLines = function () {
        var lines = self.getMatchingLines();
        var hdrText;
        if (_settings.uniqueLines)
            hdrText = "\nUnique lines with matches ({0}):";
        else
            hdrText = "\nLines with matches ({0}):";
        common.log(hdrText.format(lines.length));
        for (var l in lines) {
            common.log(lines[l]);
        }
    };
}

exports.Searcher = Searcher;